cmd/proxy/actions/app_proxy.go

```diff
@@ -18,11 +23,41 @@ func addProxyRoutes(
 	app.GET("/healthz", healthHandler)
 
 	// Download Protocol
+	// the download.Protocol and the stash.Stasher interfaces are composable
+	// in a middleware fashion. Therefore you can separate concerns
+	// by the functionality: a download.Protocol that just takes care
+	// of "go getting" things, and another Protocol that just takes care
+	// of "pooling" requests etc.
+
+	// In our case, we'd like to compose both interfaces in a particular
+	// order to ensure logical ordering of execution.
+
+	// Here's the order of an incoming request to the download.Protocol:
+
+	// 1. The downloadpool gets hit first, and manages concurrent requests
+	// 2. The downloadpool passes the request to its parent Protocol: stasher
+	// 3. The stasher Protocol checks storage first, and if storage is empty
+	// it makes a Stash request to the stash.Stasher interface.
+
+	// Once the stasher picks up an order, here's how the requests go in order:
+	// 1. The singleflight picks up the first request and latches duplicate ones.
+	// 2. The singleflight passes the stash to its parent: stashpool.
+	// 3. The stashpool manages limiting concurrent requests and passes them to stash.
+	// 4. The plain stash.New just takes a request from upstream and saves it into storage.
 	gg, err := goget.New()
 	if err != nil {
 		return err
 	}
-	p := download.New(gg, s, env.GoGetWorkers())
+	st := singleflight.New(
+		stashpool.New(
+			stash.New(gg, s),
+			env.GoGetWorkers(),
+		),
+	)
+	p := downloadpool.New(
```

Comment:
I don't entirely understand why we need the `downloadpool`. Could you expand on that? Like, what would happen if we wouldn't use it at all  and just pass `p := stasher.New(gg, s, st)` to the handler opts?

Reply:
@marpio Because we need to be able to control how many concurrent connections to storage we can open as well as how much data we copy from the storage through the proxy and back to the client.

Reply:
@marpio totally, the connections is not the issue but how much copying do you do from those connections. In other words, the memory part is the issue that I ran into :)